src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -0,0 +1,972 @@
+    std::string encryptedNote2;
+    {
+        CDataStream ss2(SER_NETWORK, PROTOCOL_VERSION);
+        ss2 << ((unsigned char) 0x00);
+        ss2 << jsdesc.ephemeralKey;
+        ss2 << jsdesc.ciphertexts[0];
+        ss2 << jsdesc.h_sig(*pzcashParams, joinSplitPubKey_);
+
+        encryptedNote1 = HexStr(ss2.begin(), ss2.end());
+    }
+    {
+        CDataStream ss2(SER_NETWORK, PROTOCOL_VERSION);
+        ss2 << ((unsigned char) 0x01);
+        ss2 << jsdesc.ephemeralKey;
+        ss2 << jsdesc.ciphertexts[1];
+        ss2 << jsdesc.h_sig(*pzcashParams, joinSplitPubKey_);
+
+        encryptedNote2 = HexStr(ss2.begin(), ss2.end());
+    }
+
+    Object obj;
+    obj.push_back(Pair("encryptednote1", encryptedNote1));
+    obj.push_back(Pair("encryptednote2", encryptedNote2));
+    obj.push_back(Pair("rawtxn", HexStr(ss.begin(), ss.end())));
+    return obj;
+}
+
+void AsyncRPCOperation_sendmany::add_taddr_outputs_to_tx() {
+
+    CMutableTransaction rawTx(tx_);
+
+    for (SendManyRecipient & r : t_outputs_) {
+        std::string outputAddress = std::get<0>(r);
+        CAmount nAmount = std::get<1>(r);
+
+        CBitcoinAddress address(outputAddress);
+        if (!address.IsValid()) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid output address, not a valid taddr.");
+        }
+
+        CScript scriptPubKey = GetScriptForDestination(address.Get());
+
+        CTxOut out(nAmount, scriptPubKey);
+        rawTx.vout.push_back(out);
+    }
+
+    tx_ = CTransaction(rawTx);
+}
+
+void AsyncRPCOperation_sendmany::add_taddr_change_output_to_tx(CAmount amount) {
+
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+
+    EnsureWalletIsUnlocked();
+    CReserveKey keyChange(pwalletMain);
+    CPubKey vchPubKey;
+    bool ret = keyChange.GetReservedKey(vchPubKey);
+    if (!ret) {
+        throw JSONRPCError(RPC_WALLET_KEYPOOL_RAN_OUT, "Could not generate a taddr to use as a change address"); // should never fail, as we just unlocked
+    }
+    CScript scriptPubKey = GetScriptForDestination(vchPubKey.GetID());
+    CTxOut out(amount, scriptPubKey);
+
+    CMutableTransaction rawTx(tx_);
+    rawTx.vout.push_back(out);
+    tx_ = CTransaction(rawTx);
+}
+
+boost::array<unsigned char, ZC_MEMO_SIZE> AsyncRPCOperation_sendmany::get_memo_from_hex_string(std::string s) {
+    boost::array<unsigned char, ZC_MEMO_SIZE> memo = {{0x00}};
+    
+    std::vector<unsigned char> rawMemo = ParseHex(s.c_str());
+
+    // If ParseHex comes across a non-hex char, it will stop but still return results so far.
+    size_t slen = s.length();
+    if (slen % 2 !=0 || (slen>0 && rawMemo.size()!=slen/2)) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Memo must be in hexadecimal format");
+    }
+    
+    if (rawMemo.size() > ZC_MEMO_SIZE) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Memo size of %d is too big, maximum allowed is %d", rawMemo.size(), ZC_MEMO_SIZE));
+    }
+    
+    // copy vector into boost array
```

Comment:
Are those not both guaranteed to be in contiguous memory, which could be copied directly?